### ISSUE TITLE
session_transaction does not work with werkzeug 2.3

### ISSUE THREAD

**Issue by azmeuk** [level12/flask-webtest#22]
`session_transaction` seems to be broken since flask/werkzeug 2.3, resulting in empty sessions. Here is a minimal reproducible example:

#### tox.ini

```ini
[tox]
skipsdist = true
envlist = py311-flask22, py311-flask23

[testenv:py311-flask22]
deps =
    flask<2.3
    werkzeug<2.3
    pytest
    flask-webtest
commands = pytest {posargs}

[testenv:py311-flask23]
deps =
    flask>=2.3
    werkzeug>=2.3
    pytest
    flask-webtest
commands = pytest {posargs}
```

#### test_foobar.py

```python
import pytest
from flask import Flask
from flask import session
from flask_webtest import TestApp


def create_app():
   app = Flask(__name__)
   app.config["SECRET_KEY"] = "secret-key"

   @app.route('/')
   def sess():
      session["foo"] = "bar"
      return 'Hello, World!'

   return app


@pytest.fixture
def app():
    return TestApp(create_app())


def test_session_session_transaction(app):
    response = app.get("/")
    response.mustcontain("Hello, World!")
    with app.session_transaction() as session:
        assert session.get("foo") == "bar"
```

If you just run `tox` you will see that the test pass with flask&werkzeug 2.2 but not with 2.3

```
E           AssertionError: assert None == 'bar'
E            +  where None = <bound method SecureCookieSession.get of <SecureCookieSession {}>>('foo')
E            +    where <bound method SecureCookieSession.get of <SecureCookieSession {}>> = <SecureCookieSession {}>.get
```

Related part of the [werkzeug 2.3 changelog](https://werkzeug.palletsprojects.com/en/2.3.x/changes/)

> Refactor parse_cookie and dump_cookie. [#2637](https://github.com/pallets/werkzeug/pull/2637)
> - parse_cookie is up to 40% faster, dump_cookie is up to 60% faster.
> - Passing bytes to parse_cookie and dump_cookie is deprecated. The dump_cookie charset parameter is deprecated.
> - dump_cookie allows domain values that do not include a dot ., and strips off a leading dot.
> - dump_cookie does not set path="/" unnecessarily by default.
>
> Refactor the test client cookie implementation. [#1060](https://github.com/pallets/werkzeug/issues/1060), [#1680](https://github.com/pallets/werkzeug/issues/1680)
> - The cookie_jar attribute is deprecated. http.cookiejar is no longer used for storage.
> - Domain and path matching is used when sending cookies in requests. The domain and path parameters default to localhost and /.
> - Added a get_cookie method to inspect cookies.
> - Cookies have decoded_key and decoded_value attributes to match what the app sees rather than the encoded values a client would see.
> - The first positional server_name parameter to set_cookie and delete_cookie is deprecated. Use the domain parameter instead.
> - Other parameters to delete_cookie besides domain, path, and value are deprecated.

**Comment by guruofgentoo** [comment]
`werkzeug` prior to version 2.3 was a bit friendlier to localhost testing with default flask settings. You'll notice that if you observe the test client's cookies with 2.3:
```
{('localhost.local', '/', 'session'): Cookie(key='session', value='eyJmb28iOiJiYXIifQ.ZQnrag.i96F7WGz5ywLd1fJGI4NBgPZwh8', decoded_key='session', decoded_value='eyJmb28iOiJiYXIifQ.ZQnrag.i96F7WGz5ywLd1fJGI4NBgPZwh8', expires=None, max_age=None, domain='localhost.local', origin_only=True, path='/', secure=False, http_only=False, same_site=None)}
```

The data is all there from the request (which you can see if you b64decode the first part of the cookie value).

When doing a session transaction, werkzeug only pulls back the cookies that match the domain. Since the test request context is setting up an environ with a plain localhost server name, the match fails, and no cookie is returned.

Adding `app.config["SERVER_NAME"] = "localhost.local"` to app setup corrects the test environ, matches the cookie, and allows the test to pass.

The change in behavior is in werkzeug's sphere, to make cookie matching more strict. Werkzeug's entire cookie mechanism was refactored, and there is admittedly a bit of translation involved in flask-webtest to keep that working with webtest. For this domain matching case, though, IMO there's nothing to work around that in flask-webtest - it's a matter of configuring the test app.

**Comment by guruofgentoo** [comment]
Yeah, this is an interesting one. The `localhost` comes from [Flask](https://github.com/pallets/flask/blob/ed5b240417414cbd0322efa95c91f759928ba154/src/flask/testing.py#L62).

But `localhost.local` doesn't come from either flask or werkzeug. Instead, it's determined in [webtest](https://github.com/Pylons/webtest/blob/main/webtest/app.py#L236). Webtest has been that way for several years, and it was the cookie handling refactor in werkzeug that made things "break."

For `flask_webtest`, that leaves us in a place where we either a) paper it over so things "just work", or b) allow the libraries to act and be configured as they appear to expect at this time. When adjusting for the cookie refactor, I chose option B. All of my active projects use `SERVER_NAME`, so I wasn't aware lacking it would trigger this behavior.

**Comment by azmeuk** [comment]
For the record, this is fixed in 0.1.6 with the last commits from @rsyring
